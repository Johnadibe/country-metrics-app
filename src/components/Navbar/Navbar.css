.navbar {
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: #0d1117;
  color: #fff;
  padding: 1.2rem 1.6rem;
}

.nav__logo__text {
  width: 3.2rem;
}

.nav__logo__text img {
  width: 100%;
}

.nav__text__center {
  font-size: 2rem;
}

.nav__logo__box .display__block {
  display: inline-flex;
}

.nav__logo__arrow {
  width: 1.5rem;
}

.nav__logo__box .display__none {
  display: none;
}

.nav__box__right {
  display: flex;
  align-items: center;
  gap: 20px;
  width: 5.6rem;
}

.nav__box__right button {
  border: 0;
  background-color: transparent;
  cursor: pointer;
}

.nav__box__right button img {
  width: 100%;
}
