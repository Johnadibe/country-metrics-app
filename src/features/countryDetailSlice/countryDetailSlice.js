import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const API = 'https://restcountries.com/v3.1/name/';

const initialState = {
  detailCountry: [],
  status: 'idle',
  error: null,
  countrySelected: '',
};

export const fetchCountryDetail = createAsyncThunk(
  'countryDetail/fetchCountryDetail',
  async (countrySelected) => {
    try {
      const response = await fetch(`${API}${countrySelected}`);
      return response.json();
    } catch (error) {
      return error.message;
    }
  },
);

export const countryDetailSlice = createSlice({
  name: 'countryDetail',
  initialState,
  reducers: {
    countrySelectedAction: (state, action) => ({
      ...state,
      countrySelected: action.payload,
    }),
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchCountryDetail.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(fetchCountryDetail.fulfilled, (state, action) => ({
        ...state,
        detailCountry: action.payload.map((countryName) => ({
          id: countryName.name.common,
          name: countryName.name.common,
          capital: countryName.capital,
          population: countryName.population,
          map: `https://raw.githubusercontent.com/djaiss/mapsicon/master/all/${countryName.cca2.toLowerCase()}/vector.svg`,
          flag: countryName.flags.png,
          area: countryName.area,
          timezones: countryName.timezones[0],
          ccn3: countryName.ccn3,
          latitude: countryName.latlng[0],
          longitude: countryName.latlng[1],
          languages: countryName.languages,
          coatOfArms: countryName.coatOfArms,
        })),
        status: 'succeeded',
      }))
      .addCase(fetchCountryDetail.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: [...state.error, action.error.message],
      }));
  },
});

export const { countrySelectedAction } = countryDetailSlice.actions;
export default countryDetailSlice.reducer;
