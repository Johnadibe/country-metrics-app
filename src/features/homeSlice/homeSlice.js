import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

// const API = 'https://restcountries.com/v3.1/region/europe';
const API = 'https://restcountries.com/v3.1/all';

const initialState = {
  countriesData: [], // Getting Data for all the countries
  status: 'idle',
  error: null,
  searchResult: '',
};

export const fetchCountries = createAsyncThunk(
  'countries/fetchCountries',
  async () => {
    try {
      const response = await fetch(API);
      return response.json();
    } catch (error) {
      return error.message;
    }
  },
);

export const homeSlice = createSlice({
  name: 'countries',
  initialState,
  reducers: {
    search: (state, action) => ({
      ...state, searchResult: action.payload,
    }),
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchCountries.pending, (state) => ({
        ...state, status: 'loading',
      }))
      .addCase(fetchCountries.fulfilled, (state, action) => ({
        ...state,
        status: 'succeeded',
        countriesData: action.payload.map((country) => ({
          id: country.cca3,
          name: country.name.common,
          capital: country.capital,
          population: country.population,
          map: `https://raw.githubusercontent.com/djaiss/mapsicon/master/all/${country.cca2.toLowerCase()}/vector.svg`,
        })),
      }))
      .addCase(fetchCountries.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: [...state.error, action.error.message],
      }));
  },
});

export const { search } = homeSlice.actions;
export default homeSlice.reducer;
