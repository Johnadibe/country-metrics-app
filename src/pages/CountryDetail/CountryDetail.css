.detail__body {
  background-color: #161b22;
}

.detail__header__box {
  display: grid;
  grid-template-columns: 1fr 1fr;
  height: 20rem;
  align-items: center;
  justify-content: space-between;
  padding: 0 1.6rem;
  background-color: #161b22;
}

.detail__header__box__map {
  width: 100%;
}

.detail__header__box__map img {
  width: 80%;
}

.detail__header__box__text {
  text-align: right;
  font-size: 1.8rem;
  color: #fff;
}

.country__breakdown__text {
  font-size: 1.6rem;
  color: #fff;
  background-color: #0d1117;
  padding: 1rem 1.6rem;
}

.detail__breakdown__box {
  display: flex;
  flex-direction: column;
}

.detail__breakdown__box__text {
  display: grid;
  grid-template-columns: 1fr 10rem 2rem;
  align-items: center;
  padding: 2.5rem 1rem 2.5rem 0.5rem;
  column-gap: 3.5rem;
  background-color: #161b22;
}

.detail__breakdown__box__text .country__flag {
  width: 5rem;
}

.detail__breakdown__box__text h3,
p {
  font-size: 1.8rem;
  color: #fff;
}

.detail__breakdown__box__text button {
  border: 0;
  background-color: transparent;
  cursor: pointer;
}

.color__dark {
  background-color: #0d1117;
}

.error__back {
  text-align: center;
  color: #fff;
  padding: 10rem 0 10rem 0;
}
