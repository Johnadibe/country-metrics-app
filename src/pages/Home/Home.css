.home__body {
  background-color: #161B22;
  height: 100vh;
}

.country__grid {
  display: grid;
  grid-template-columns: 1fr 1fr;
  align-items: center;
  padding: 0 1.6rem;
  background-color: #161B22;
}

.card__flex-column {
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 20rem;
  position: relative;
}

.country__arrow__link {
  width: 2rem;
  border: 0;
  background-color: transparent;
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  margin-top: 1rem;
}

.country__arrow__link img {
  width: 100%;
}

.card__text {
  text-align: end;
  color: #fff;
}

.card__text p {
  font-size: 1.6rem;
  font-weight: 400;
  line-height: 2.4rem;
}

.grid__image {
  width: 50%;
  margin: auto;
}

.grid__image img {
  width: 100%;
}

.outline-line {
  border-left: 2px solid #7a668d;
}

.card_text_none {
  grid-column: 1 / -1;
  text-align: center;
  color: #fff;
  padding: 2rem 0 2rem 0;
}

