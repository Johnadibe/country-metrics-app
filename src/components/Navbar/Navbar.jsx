import React from 'react';
import { Link, useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';
import micro from '../../assests/svg/microphone-solid.svg';
import gear from '../../assests/svg/gear-solid.svg';
import arrrowLeft from '../../assests/svg/angle-left-solid.svg';
import logo from '../../assests/image/logoimg.jpg';
import './Navbar.css';

const Navbar = () => {
  const countrySelected = useSelector((state) => state.countryDetail.countrySelected);

  const usePathname = () => {
    const location = useLocation();
    return location.pathname;
  };
  return (
    <>
      <nav className="navbar">
        <div className="nav__logo__box">
          <Link to="/" className={usePathname() === '/countrydetail' ? 'display__block' : 'display__none'}>
            <img
              src={arrrowLeft}
              alt="Angle Left Arrow"
              className="nav__logo__arrow"
            />
          </Link>
          <div className="nav__logo__text">
            {usePathname() === '/countrydetail' ? '' : <img className="logo_img" src={logo} alt="Logo" /> }
          </div>
        </div>

        <span className="nav__text__center">
          {usePathname() === '/countrydetail' ? `${countrySelected} Details` : 'Europe Population'}
        </span>

        <div className="nav__box__right">
          <button type="button">
            <img src={micro} alt="Microphone" />
          </button>
          <button type="button">
            <img src={gear} alt="Gear" />
          </button>
        </div>
      </nav>
    </>
  );
};

export default Navbar;
