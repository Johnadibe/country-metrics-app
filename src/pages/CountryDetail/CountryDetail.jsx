import React from 'react';
import { useSelector } from 'react-redux';
import rightArrow from '../../assests/svg/circle-arrow-right-solid.svg';
import Footer from '../../components/Footer/Footer';
import './CountryDetail.css';

const CountryDetail = () => {
  const country = useSelector((state) => state.countryDetail.detailCountry);
  return (
    <div className="detail__body">
      {
        country.length > 0 ? (
          <div>
            <div className="detail__header__box">
              <div className="detail__header__box__map">
                <img src={country[0].map} alt="" />
              </div>
              <div className="detail__header__box__text">
                <h1>{country[0].name}</h1>
                <p>
                  {country[0].population}
                  {' '}
                  Population
                </p>
              </div>
            </div>
            <h3 className="country__breakdown__text">COUNTRY BREAKDOWN - 2023</h3>
            <div className="detail__breakdown__box">
              <div className="detail__breakdown__box__text">
                <h3>Area</h3>
                <p>
                  {country[0].area}
                  {' '}
                  km
                </p>
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
              <div className="detail__breakdown__box__text color__dark">
                <h3>National Flag</h3>
                <img className="country__flag" src={country[0].flag} alt={`${country[0].name} Flag`} />
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
              <div className="detail__breakdown__box__text">
                <h3>Longitude</h3>
                <p>{country[0].longitude}</p>
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
              <div className="detail__breakdown__box__text color__dark">
                <h3>Latitude</h3>
                <p>{country[0].latitude}</p>
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
              <div className="detail__breakdown__box__text">
                <h3>Cellular Communication Network</h3>
                <p>{country[0].ccn3}</p>
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
              <div className="detail__breakdown__box__text color__dark">
                <h3>Timezones</h3>
                <p>{country[0].timezones}</p>
                <button type="button">
                  <img src={rightArrow} alt="Right Arrow" />
                </button>
              </div>
            </div>
          </div>
        ) : (<h1 className="error__back">Oops! Please press back </h1>)
}
      <Footer />
    </div>
  );
};

export default CountryDetail;
