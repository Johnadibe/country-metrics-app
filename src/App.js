import React from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
// import Navbar from './components/Navbar/Navbar';
import Home from './pages/Home/Home';
import CountryDetail from './pages/CountryDetail/CountryDetail';
import './App.css';

function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/countrydetail/:name" element={<CountryDetail />} />
      </Routes>
    </BrowserRouter>
  );
}

export default App;
