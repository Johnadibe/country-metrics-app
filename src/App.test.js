import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import App from "./App";
import store from "./redux/store/store";
import Navbar from "./components/Navbar/Navbar";
import Home from "./pages/Home/Home";
import CountryDetail from "./pages/CountryDetail/CountryDetail";
import { fetchCountries, search } from "./features/homeSlice/homeSlice";
import { fetchCountryDetail } from "./features/countryDetailSlice/countryDetailSlice";

it("renders App without crashing", () => {
    const root = renderer.create(
        <Provider store={store}>
            <App />
        </Provider>,
    ).toJSON();
    expect(root).toMatchSnapshot();
});

it("renders Navbar without crashing", () => {
    const root = renderer.create( 
        <BrowserRouter>
        <Provider store={store}>
            <Navbar />
        </Provider>,
        </BrowserRouter>,
    ).toJSON();
    expect(root).toMatchSnapshot();
});

it("renders Home without crashing", () => {
    const root = renderer.create(
        <BrowserRouter>
        <Provider store={store}>
            <Home />
        </Provider>,
        </BrowserRouter>,
    ).toJSON();
    expect(root).toMatchSnapshot();
});

it("renders CountryDetail without crashing", () => {
    const root = renderer.create(
        <BrowserRouter>
        <Provider store={store}>
            <CountryDetail />
        </Provider>,
        </BrowserRouter>,
    ).toJSON();
    expect(root).toMatchSnapshot();
});

// HomeSlice and Reducer Check

describe("Country Home redux state test", () => {
    it('Should initially Country store to an empty Array', () => {
        const state = store.getState().country;
        expect(state.countriesData).toEqual([]);
      });

      it('should Country payload send correct', () => {
        const expectedPayload = { payload: 'Serbia', type: 'countries/search' };
        const actualPayload = search('Serbia');
        expect(actualPayload).toEqual(expectedPayload);
      });

      it('fetch Country data from API', async () => {
        const url = 'https://restcountries.com/v3.1/region/europe';
        const axiosSpy = jest.spyOn(axios, 'get');
        jest.setTimeout(90000);
        const dispatchSpy = jest.fn();
    
        await fetchCountries(url)(dispatchSpy);
    
        expect(axiosSpy).toHaveBeenCalledWith(url);
      });
});

// country Deatai slicer and reducer check

describe('country Detail redux state tests', () => {
  it('Should initially Country detail store to an empty Array', () => {
    const state = store.getState().countryDetail;
    expect(state.detailCountry).toEqual([]);
  });
  it('fetch Country data from API', async () => {
    const url = 'https://restcountries.com/v3.1/name/serbia';
    const axiosSpy = jest.spyOn(axios, 'get');
    jest.setTimeout(90000);
    const dispatchSpy = jest.fn();

    await fetchCountryDetail('Serbia', url)(dispatchSpy);

    expect(axiosSpy).toHaveBeenCalledWith(url);
  });
});