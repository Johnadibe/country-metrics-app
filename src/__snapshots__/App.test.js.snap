// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App without crashing 1`] = `
Array [
  <nav
    className="navbar"
  >
    <div
      className="nav__logo__box"
    >
      <a
        className="display__none"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="Angle Left Arrow"
          className="nav__logo__arrow"
          src="angle-left-solid.svg"
        />
      </a>
      <div
        className="nav__logo__text"
      >
        <img
          alt="Logo"
          className="logo_img"
          src="logoimg.jpg"
        />
      </div>
    </div>
    <span
      className="nav__text__center"
    >
      Europe Population
    </span>
    <div
      className="nav__box__right"
    >
      <button
        type="button"
      >
        <img
          alt="Microphone"
          src="microphone-solid.svg"
        />
      </button>
      <button
        type="button"
      >
        <img
          alt="Gear"
          src="gear-solid.svg"
        />
      </button>
    </div>
  </nav>,
  <div
    className="home__body"
  >
    <div
      className="europe__population"
    >
      <div
        className="europe__population__img__box"
      >
        <img
          alt="Europe Map"
          className="europe__img"
          src="https://svgsilh.com/svg/151588.svg"
        />
      </div>
      <div
        className="center-line"
      />
      <div
        className="europe__population__text"
      >
        <h1>
          Europe
        </h1>
        <p>
          Loading...
           
          Population
        </p>
      </div>
    </div>
    <div
      className="search__box"
    >
      <img
        alt=""
        className="search__icon"
        src="magnifying-glass-solid.svg"
      />
      <input
        className="search__input"
        onChange={[Function]}
        placeholder="Search for a country..."
        type="text"
      />
    </div>
    <div
      className="country__grid"
    >
      <h1
        className="card_text_none"
      >
        OOps! No Country Found. Please Search Again
      </h1>
    </div>
    <div>
      <h3
        className="footer"
      >
        @Copyright. 2023; Built by John C. Adibe
      </h3>
    </div>
  </div>,
]
`;

exports[`renders CountryDetail without crashing 1`] = `
Array [
  <div
    className="detail__body"
  >
    <h1
      className="error__back"
    >
      Oops! Please press back 
    </h1>
    <div>
      <h3
        className="footer"
      >
        @Copyright. 2023; Built by John C. Adibe
      </h3>
    </div>
  </div>,
  ",",
]
`;

exports[`renders Home without crashing 1`] = `
Array [
  <div
    className="home__body"
  >
    <div
      className="europe__population"
    >
      <div
        className="europe__population__img__box"
      >
        <img
          alt="Europe Map"
          className="europe__img"
          src="https://svgsilh.com/svg/151588.svg"
        />
      </div>
      <div
        className="center-line"
      />
      <div
        className="europe__population__text"
      >
        <h1>
          Europe
        </h1>
        <p>
          Loading...
           
          Population
        </p>
      </div>
    </div>
    <div
      className="search__box"
    >
      <img
        alt=""
        className="search__icon"
        src="magnifying-glass-solid.svg"
      />
      <input
        className="search__input"
        onChange={[Function]}
        placeholder="Search for a country..."
        type="text"
      />
    </div>
    <div
      className="country__grid"
    >
      <h1
        className="card_text_none"
      >
        OOps! No Country Found. Please Search Again
      </h1>
    </div>
    <div>
      <h3
        className="footer"
      >
        @Copyright. 2023; Built by John C. Adibe
      </h3>
    </div>
  </div>,
  ",",
]
`;

exports[`renders Navbar without crashing 1`] = `
Array [
  <nav
    className="navbar"
  >
    <div
      className="nav__logo__box"
    >
      <a
        className="display__none"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="Angle Left Arrow"
          className="nav__logo__arrow"
          src="angle-left-solid.svg"
        />
      </a>
      <div
        className="nav__logo__text"
      >
        <img
          alt="Logo"
          className="logo_img"
          src="logoimg.jpg"
        />
      </div>
    </div>
    <span
      className="nav__text__center"
    >
      Europe Population
    </span>
    <div
      className="nav__box__right"
    >
      <button
        type="button"
      >
        <img
          alt="Microphone"
          src="microphone-solid.svg"
        />
      </button>
      <button
        type="button"
      >
        <img
          alt="Gear"
          src="gear-solid.svg"
        />
      </button>
    </div>
  </nav>,
  ",",
]
`;
